### ISSUE TITLE
Supervisord monitoring

### ISSUE THREAD

**Issue by ilyam8** [netdata/netdata#4852]
We are willing to support monitoring ***[Supervisord](http://supervisord.org/)*** with netdata.

To get started, we need a user that will act as a [sponsor](https://docs.netdata.cloud/contributing/#sponsor-a-part-of-netdata) during the implementation.
He/she will assist us with the following:

  - Specify the metrics required and how they should be presented in charts.
  - Specify the alarms that would make sense for each metric.
  - When the implementation passes QA, test the implementation in production.
  - Use the charts and alarms in their day to day work and provide additional feedback after the collector is delivered.

We would also appreciate getting votes :+1: for this issue from the community, to understand the interest in the particular collector.

**Comment by cakrit** [comment]
We have some interest here, but no sponsor. Can someone help us get this started?

**Comment by ekexcello** [comment]
Hello @ilyam8 

As we are using supervisord and netdata in everyday work, we also require plugin to monitor it.
From our point, metrics would look as follows:
* list of configured supervised jobs
* number of running processes for every job
* number of *not* running processes per job
* status of each process (running/starting/failed etc)
* uptime of each process (combined to 1 chart for every configured job)

At the moment, alarms required are:
* for process entered FAIL or BACKOFF state
* process exited with unexpected status
* process uptime is less then configured threshold

Also, please note the fact that complicates this monitoring, namely, access level as many installations run supervisor as root, so access for netdata user to supervisor socket or interface is a point of configuration.

**Comment by ilyam8** [comment]
@ekexcello short guide how to setup a lab (docker) with supervisord with several jobs/processes is very welcome!

**Comment by ekexcello** [comment]
Hello @ilyam8 
Sorry for the delay.
You can find set of files to create docker running supervisor and testing jobs to experement [here](https://github.com/ekexcello/supervisor_laba). This is common case with supervisor started as root.
Please feel free to ask for detail if required.

**Comment by ilyam8** [comment]
@ekexcello thx, i setup the lab and get processes info metrics ([`supervisor.getAllProcessInfo`](http://supervisord.org/api.html#supervisor.rpcinterface.SupervisorNamespaceRPCInterface.getAllProcessInfo)):

```go
{Name:00 Group:proc1 Start:1607884769 Stop:1607884771 Now:1607887222 State:200 StateName:FATAL SpawnErr: ExitStatus:0 StdoutLogfile: StderrLogfile: PID:0}
{Name:00 Group:proc2 Start:1607887220 Stop:1607887220 Now:1607887222 State:20 StateName:RUNNING SpawnErr: ExitStatus:0 StdoutLogfile: StderrLogfile: PID:5192}
{Name:01 Group:proc2 Start:1607887216 Stop:1607887216 Now:1607887222 State:20 StateName:RUNNING SpawnErr: ExitStatus:0 StdoutLogfile: StderrLogfile: PID:5186}
{Name:02 Group:proc2 Start:1607887219 Stop:1607887219 Now:1607887222 State:20 StateName:RUNNING SpawnErr: ExitStatus:0 StdoutLogfile: StderrLogfile: PID:5190}
{Name:00 Group:proc3 Start:1607884756 Stop:0 Now:1607887222 State:20 StateName:RUNNING SpawnErr: ExitStatus:0 StdoutLogfile: StderrLogfile: PID:16}
```

> list of configured supervised jobs
number of running processes for every job
number of not running processes per job
status of each process (running/starting/failed etc)
uptime of each process (combined to 1 chart for every configured job)

About charts, do you mean a subsection per a job (process’ group), something like this: (?)

```yaml
supervisord:
  - proc1:
      process state chart:
        - 00
      process exit status chart:
        - 00
      process uptime chart:
        - 00
  - proc2:
      process state chart:
        - 00
        - 01
        - 02
      process exit status chart:
        - 00
        - 01
        - 02
      process uptime chart:
        - 00
        - 01
        - 02
  - proc3:
    ...
```

**Comment by ekexcello** [comment]
Hello Ilya, 

Your proposed chart structure looks perfect! 
It would be nice to have subsection per job if possible. However, on some setups number of supervised jobs can be high (over 20 groups, for example), thus it would be also good to have configuration option to enable or disable charts of certain type (for example, someone won't need 'exit statuses chart' or some other chart and would like to hide it to scroll less).
Also, I discussed proposed chart list with our team, they suggested additional summarizing charts and some additional helpful alarms, namely,
summarizing charts:
* process group uptime (per group, displays summarized uptime of all processes within the group)
* process group downtime (per group, displays summarized time when no process was running within the group)
* global uptime: single chart - summarized uptime of all groups (mean, summarized time when at least one process in each group was up since supervisord startup). If it will be complex to track across netdata restarts, this can be treated as 'since last supervisor or netdata restart'.
* global downtime: single chart - summarized downtime of all process groups (summarized time when all the processes of one or more groups were completely down). With same remark as previous point.
* I'd like to request group summary probably, in the top of each group.

Considering this, final chart structure looks in following way:
```
supervisord:
  - proc1 (name):
      process group summary:
        - number of running processes in group at the moment
        - number of non-running processes in group at the moment
      average process uptime
        - average uptime of processes within the group.
      overall process group uptime chart:
        - summarized uptime of all processes within the group
      overall process group downtime chart:
        - summarizing time when no process was running in a group
      process state chart:
        - 00 <current_state>
      process exit status chart:
        - 00 <exit_status>
      process uptime chart:
        - 00 <uptime_seconds>
  - proc2 (name):
      process group summary:
        - number of running processes in group at the moment
        - number of non-running processes in group at the moment
      average process uptime
        - average uptime of processes within the group.
      overall process group uptime chart:
        - summarized uptime of all processes within the group
      overall process group downtime chart:
        - summarizing time when no process was running in a group
      process state chart:
        - 00 <current_state>
        - 01 <current_state>
        - 02 <current_state>
      process exit status chart:
        - 00 <exit_status>
        - 01 <exit_status>
        - 02 <exit_status>
      process uptime chart:
        - 00 <uptime_seconds>
        - 01 <uptime_seconds>
        - 02 <uptime_seconds>
  - proc3:
  ...
  - global uptime:
        - Single chart. Summarized uptime of all process groups
  - global downtime:
        - Single chart. Summarized downtime of all process groups
```
As for alarms, slightly adjusted list looks as follows:
* process entered FAIL or BACKOFF state
* process exited with unexpected status
* process uptime is less then configured threshold
* single process is running too short (threshold can vary across groups)
* single process is running too long (threshold can vary across groups)
* process group is down too long

**Comment by ilyam8** [comment]
@ekexcello thx for the detailed response, looks very good!

**Comment by ilyam8** [comment]
@hamedbrd if you are still willing to work on this - there a guide how to setup a lab env.

@ekexcello i simplified your setup (removed Netdata and added http server)

### Lab setup/teardown

```sh
# download
git clone https://github.com/ilyam8/supervisor_laba
cd supervisor_laba

# build and start in detached mode
docker-compose up --build --detach

# stop
docker-compose stop

# cleanup
docker rmi $(docker images -a --filter=dangling=true -q)
docker rmi supervisor_laba
```

### Go program to collect jobs statuses

I use [same xmlrpc library as node exporter uses](https://github.com/prometheus/node_exporter/blob/0ce1f39c64a0a37d5aa81da8006661920b097e82/collector/supervisord.go#L28) - not sure it is the best choice 🤷‍♂️ , but it works and it is not important what we use now (can change later).

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net"
	"net/http"
	"net/url"
	"time"

	"github.com/mattn/go-xmlrpc"
)

type processInfo struct {
	Name          string `xmlrpc:"name"`
	Group         string `xmlrpc:"group"`
	Start         int    `xmlrpc:"start"`
	Stop          int    `xmlrpc:"stop"`
	Now           int    `xmlrpc:"now"`
	State         int    `xmlrpc:"state"`
	StateName     string `xmlrpc:"statename"`
	SpawnErr      string `xmlrpc:"spanerr"`
	ExitStatus    int    `xmlrpc:"exitstatus"`
	StdoutLogfile string `xmlrcp:"stdout_logfile"`
	StderrLogfile string `xmlrcp:"stderr_logfile"`
	PID           int    `xmlrpc:"pid"`
}

func newXMLRPCClient(serverURL string) *xmlrpc.Client {
	u, _ := url.Parse(serverURL)

	switch u.Scheme {
	case "http":
		return xmlrpc.NewClient(serverURL)
	case "unix":
		client := xmlrpc.NewClient("http://unix/RPC2")
		client.HttpClient.Transport = &http.Transport{
			DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
				d := net.Dialer{Timeout: 10 * time.Second}
				return d.DialContext(ctx, "unix", u.Path)
			},
		}
		return client
	default:
		return nil
	}
}

func main() {
	const (
		unixHTTPServerURL = "unix:///run/supervisor.sock"
		inetHTTPServerURL = "http://localhost:9001/RPC2"
	)

	rpcClient := newXMLRPCClient(inetHTTPServerURL)
	defer rpcClient.HttpClient.CloseIdleConnections()

	resp, err := rpcClient.Call("supervisor.getAllProcessInfo")
	if err != nil {
		log.Println(err)
		return
	}

	res, ok := resp.(xmlrpc.Array)
	if !ok {
		return
	}

	for _, p := range res {
		s, ok := p.(xmlrpc.Struct)
		if !ok {
			continue
		}

		var info processInfo
		for k, v := range s {
			switch k {
			case "name":
				info.Name = v.(string)
			case "group":
				info.Group = v.(string)
			case "start":
				info.Start = v.(int)
			case "stop":
				info.Stop = v.(int)
			case "now":
				info.Now = v.(int)
			case "state":
				info.State = v.(int)
			case "statename":
				info.StateName = v.(string)
			case "exitstatus":
				info.ExitStatus = v.(int)
			case "pid":
				info.PID = v.(int)
			}
		}
		fmt.Printf("%+v\n", info)
	}
}
```

### Get jobs statuses from supervisord via built-in HTTP server

```sh
go build <program_name>.go
./<program_name>
```

**Comment by ilyam8** [comment]
@ekexcello i am far from being a docker specialist 🙃

I tried to get data collection via unix socket working from the host, but failed 😞

I did same as you did with logs directory and got `supervisor.sock` on the host - result is `connection refused` on dial. It works from inside the container tho.

Not a big problem, just sharing my experience. Workaround is `docker cp` binary inside the container and `docker exec -it`.

Fortunately, Supervisor exposes data via built-in HTTP server, so we can use it developing the collector.
